components/events/tests/event_pipeline.test.cpp

```diff
@@ -0,0 +1,313 @@
+		}
+	}
+
+  private:
+	std::function<void()>           m_update{nullptr};
+	std::function<void(EventBus &)> m_attach{nullptr};
+};
+
+struct TestEvent
+{};
+
+// tracks the order of pipeline execution
+class TestStage : public TypedEventPipelineStage<TestEvent>
+{
+  public:
+	TestStage(uint32_t &current_value, uint32_t expected_value) :
+	    m_current_value{current_value}, m_expected_value{expected_value}
+	{}
+
+	virtual ~TestStage() = default;
+
+	virtual void emit(EventBus &bus) override
+	{
+		auto sender = bus.request_sender<TestEvent>();
+		sender->push(TestEvent{});
+
+		REQUIRE(m_current_value == m_expected_value);
+
+		m_current_value++;
+	}
+
+  protected:
+	uint32_t &m_current_value;
+	uint32_t  m_expected_value;
+};
+
+class TestEventPipeline final : public EventPipeline
+{
+  public:
+	TestEventPipeline()          = default;
+	virtual ~TestEventPipeline() = default;
+
+	inline size_t once_count() const
+	{
+		return m_once_stages.size();
+	}
+
+	inline size_t always_count() const
+	{
+		return m_stages.size();
+	}
+
+	inline size_t total_stage_count() const
+	{
+		return once_count() + always_count();
+	}
+};
+
+TEST_CASE("register once stage", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.once_count() == 0);
+
+	pipeline.add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.once_count() == 1);
+}
+
+TEST_CASE("register then stage", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.always_count() == 0);
+
+	pipeline.add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.always_count() == 1);
+}
+
+TEST_CASE("register multiple stages", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	REQUIRE(pipeline.always_count() == 0);
+	REQUIRE(pipeline.once_count() == 0);
+
+	pipeline
+	    .add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_once(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>())
+	    .add_always(std::make_unique<TypedEventPipelineStage<TestEvent>>());
+
+	REQUIRE(pipeline.always_count() == 3);
+	REQUIRE(pipeline.once_count() == 2);
+	REQUIRE(pipeline.total_stage_count() == 5);
+}
+
+TEST_CASE("stages are executed in the correct order", "[events]")
+{
+	TestEventPipeline pipeline{};
+
+	uint32_t execution_index{0};
+
+	pipeline
+	    .add_once(std::make_unique<TestStage>(execution_index, 0))
+	    .add_once(std::make_unique<TestStage>(execution_index, 1))
+	    .add_always(std::make_unique<TestStage>(execution_index, 2))
+	    .add_always(std::make_unique<TestStage>(execution_index, 3))
+	    .add_always(std::make_unique<TestStage>(execution_index, 4));
+
+	// assertions inside TestStage
+	pipeline.process();
+
+	// on a second pass we expect only the "then" stages to be processed
+	execution_index = 2;
+
+	pipeline.process();
```

Comment:
I don't see how the correct order is verified. Could you please elaborate a bit?